### PR TITLE
NIFI-12887 Additional ways for handling Strings as binary in PutDatabaseRecord

### PR DESCRIPTION
**Pull request by tpalfy**
Add option to treat Strings as hexadecimal character sequences or base64-encoded binary data when inserting into a binary type column.

# Summary

[NIFI-12887](https://issues.apache.org/jira/browse/NIFI-12887)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:802]: The standard Java [Base64](https://docs.oracle.com/en/java/javase/21/docs//api/java.base/java/util/Base64.html) class should be used instead of the the Google class.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:804]: The standard Java [HexFormat](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/HexFormat.html) class should be used instead of the Google class.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:263]: Recommend adjusting the description for clarity to indicate decoding.
```suggestion
            .description("The format to be applied when decoding string values to binary.")
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:250]: Recommend using `Hexadecimal` to align with other options.
```suggestion
            "Hexadecimal",
            "Hexadecimal",
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:256]: Recommend using `Base64` to align with other options.
```suggestion
            "Base64",
            "Base64",
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:804]: Java 8 support is not a requirement for the main branch, and HexFormat was introduced in Java 17, so this should be the direction for future integration. If the goal is to support this capability for the 1.x branch, then the feature could be backported and the Google library could be used. Going forward, however, we should make use of available Java capabilities as opposed to placing more dependencies on Google Guava or other libraries.

Comment (tpalfy) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:804]: Yes, I'm thinking about having this change on the 1.x branch.

So not sure what is your suggestion. If we want it on 1.x branch is it okay to use the Google library?

Or you suggest it's worth the additional work to use vanilla Java on main and instead of simply doing a cherry-pick, the committer should do the code rewrite during backport to 1.x?

**Comment by pvillard31** [comment]
@tpalfy - I think the easiest is to keep this PR for support/1.x, and open another PR targeted at main that does not use Google library. I think it's safer compared to letting the merger do the changes. Happy to merge once we're on the same page.

**Comment by tpalfy** [comment]
> @tpalfy - I think the easiest is to keep this PR for support/1.x, and open another PR targeted at main that does not use Google library. I think it's safer compared to letting the merger do the changes. Happy to merge once we're on the same page.

This PR is based on and is open against `main` already. So I'll need to change this one and open a new one against 1.x.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for the update @tpalfy, the change to the standard Java classes looks good, the only thing remaining appears to be removing the unused import for `com.google.common.io.BaseEncoding`.

**Comment by pvillard31** [comment]
Yeah the option of opening another PR for 1.x is also OK :) As a FYI, from the Github UI, on the PR, you can change the branch against which the PR is submitted, that is why I suggested to use this one against 1.x and open a new one with the changes. But all good, the end result is the same ;)

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for making the adjustments @tpalfy.

The last item to address it the `required()` statue for the new property. As mentioned in the comment, it can be set to `true` since the default value provided maintains compatibility with existing configuration.

Other than that, the changes look good.

**Comment by tpalfy** [comment]
> The last item to address it the `required()` status for the new property. As mentioned in the comment, it can be set to `true` since the default value provided maintains compatibility with existing configuration.

Technically we could make it required, but we don't automatically do that in such cases. When a property is essential on a conceptual level, we make it required but leave it optional otherwise, even if it has a default value.

This is more about UX but it makes sense. This can't be considered a conceptually essential parameter, like 'Table Name' for example. Let's leave it optional.

**Comment by exceptionfactory** [comment]
> > The last item to address it the `required()` status for the new property. As mentioned in the comment, it can be set to `true` since the default value provided maintains compatibility with existing configuration.
> 
> Technically we could make it required, but we don't automatically do that in such cases. When a property is essential on a conceptual level, we make it required but leave it optional otherwise, even if it has a default value.
> 
> This is more about UX but it makes sense. This can't be considered a conceptually essential parameter, like 'Table Name' for example. Let's leave it optional.

Thanks for the reply.

Although use of the property is optional in one respect, in that it does not apply if the record does not relate to a binary column, it is required in terms of understanding the default behavior.

Right now, the default behavior is implicit, always using UTF-8 decoding. From another angle, making the property optional leaves the behavior ambiguous, because the property value could be an empty string. In that case, it could fall back to UTF-8, but that is less clear than marking the property as required.

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the adjustments @tpalfy, the latest version looks good. +1 merging

**Pull request closed without merging**
